### PR TITLE
Configure Cirrus CI

### PR DESCRIPTION
**Pull request by fkorotkov**
@mmcc007 wanted to compare Cirrus with the current setup so I've tried to migrate. While testing I've noticed that times for different shard varies a lot for one run to another. I've checked a few Travis runs and saw the same pattern. Test run times unfortunately are not predictable. So I'm not sure how do the comparison 😔

But the config is 5 times smaller 😅 

You can check a full build here: https://cirrus-ci.com/build/5102555485962240

Just wanted to share the findings. Please let me know if you have questions. 🙌

### CODE REVIEW

**Comment by mmcc007** [comment]
Fantastic.. I've been planning on testing on Cirrus CI. So this is great.

The idea of the integration testing is to test the app(s) on iOS and android. So we run the android emulator on a linux box, and the iOS simulator on a macOS box. 

The command 'flutter devices' should find one emulator on a linux box, and one simulator on a macOS box. The 'flutter driver' will blindly run the integration test on whatever 'device' is available.

Yes, the build times are all over the place and timeouts are happening. So these new build results will help find where to focus to get more consistent builds.

BTW: does the macOS box have the simulators pre-installed (couldn't tell from a quick glance)?

**Comment by fkorotkov** [comment]
@mmcc007 yeah, `mojave-flutter` image has everything pre-installed including the latest stable Flutter. You can check the [packer file for it](https://github.com/cirruslabs/osx-images/blob/master/templates/mojave-flutter.json).

I think you might be interested in the following line as well:

```yaml
skip: '!changesInclude(".cirrus.yml", "$SHARD/*", "$SHARD/**/*")'
```

which will make CI to run for a shard only if there were changes in that directory. This should drastically improve feedback loop on PRs. 🙌

**Comment by mmcc007** [comment]
I set-up an experiment using a stripped down version of your .cirrus.yml:
https://github.com/mmcc007/flutter_arch_samples_cirrus_ci/blob/4c05306dbcbbdb782296600f979c8afad3468c3f/.cirrus.yml

It ran the unit tests and the integration test on iOS and android for vanilla project:
https://cirrus-ci.com/build/5164639909314560
(other than the failed iOS job -- due to test timeout -- the stripped down cirrus.yml looks correct)

Basically need to repeat that process for the remainder of the projects. Is this something that sharding does, or should use yaml anchors and aliases, or some other way?

**Comment by fkorotkov** [comment]
@mmcc007, you can put something like:

```yaml
env:
  matrix:
    STEP: 1
    STEP: 2
    STEP: 3
    STEP: 4
```

inside a `task` to duplicate it 4 times for example.

BTW you probably want to remove `skip` in your experiment to always run the tests.

**Comment by mmcc007** [comment]
That worked.. thnx!

Added the additional jobs and made other jobs dependent on the unit test job. Is there a way to fast fail the unit test job (rather than making all other jobs depend on this job)?
https://github.com/mmcc007/flutter_arch_samples_cirrus_ci/blob/fac8375a42c3863620dbca2dee65edc3eb5e1cac/.cirrus.yml
https://cirrus-ci.com/build/5386954555260928

**Comment by mmcc007** [comment]
Other than the that, it looks great to me!
If you want to also add a badge to the README and remove the nohup (doesn't seem to make a difference), it should be good to go.

**Comment by mmcc007** [comment]
@brianegan do you want to enable cirrus-ci for this repo?

**Comment by fkorotkov** [comment]
@mmcc007 without `nohup` I've experienced this issue: https://github.com/dart-lang/coverage/issues/249 🤷‍♂️

There is no fail fast feature at the moment but feel free to [create a new feature request](https://github.com/cirruslabs/cirrus-ci-docs/issues/new?assignees=&labels=feature&template=feature_request.md) 🙌

I'll add a badge once the app is enabled so we can see the checks on this PR.

**Comment by brianegan** [comment]
Hey all! I'm happy to have cirrus support. One question: Should we move away from travis to Cirrus?

For PRs, it might be a bit of a nightmare if two separate CIs run and report errors, maybe even different errors.

**Comment by fkorotkov** [comment]
@brianegan that's your call. 😅I'd suggest to run in parallel for some time and then decide plus it will give everyone time to rebase. I'll be happy to assist if any issues will occur. 🙌

**Comment by mmcc007** [comment]
@brianegan Ah, good point...

May be possible to disable one or the other from building during a PR (I think Cirrus has this feature and believe Travis to have it also?).

Running builds on both (say in a fork before PR) may be useful for narrowing down on build stability and speed issues (if both are enabled in the forked repo). 

Other than that, it's not necessary to have both but is an option (especially if stability issue is resolved).

I can continue experimenting in a fork (or cloned repo) for build stability and speed, so I'm ok with whichever direction.

**Comment by mmcc007** [comment]
Our first complete cirrus build (without timeouts or restarts -- what luck!) just completed in 36.14 minutes:
https://cirrus-ci.com/build/5386954555260928

This compares to travis builds of about 2 hours:
https://travis-ci.org/brianegan/flutter_architecture_samples/builds/506363464

9(8 linux and 1 macOS)? runners on cirrus, compare to 5 runners(with 2 runner limit on macOS) on travis.

So, if stability is comparable on both (Cirrus uses runners that can be pre-empted by Google?, not sure about Travis), then Cirrus has a definite speed advantage.

**Comment by mmcc007** [comment]
@fkorotkov If a runner is pre-empted, can we tell from the build report?

**Comment by fkorotkov** [comment]
@mmcc007 preempted tasks are re-started automatically on VMs that can't be preempted (more expensive ones) so it should be not too distracting (but you'll see a badge if was an automatic rerun and a link to the preempted task).

From users feedback Cirrus behaves more stable performance wise. For example [Puppeteer project](https://github.com/GoogleChrome/puppeteer) struggled pretty frequently from tests timing out prior moving to Cirrus. So maybe it's not a luck. 😉

Also worth to notice that Cirrus has a per-user limitation where Travis enforces it on an organization level [as far as I know](https://stackoverflow.com/a/34809448/1390817).

**Comment by mmcc007** [comment]
@fkorotkov Can a Cirrus build be disabled on a PR?

**Comment by fkorotkov** [comment]
@mmcc007 I can add the following as well:

```yaml
only_if: $CIRRUS_TAG == ''
```

Do you want me to add it?

**Comment by fkorotkov** [comment]
I see you are working of my config. Feel free to close this PR in favor of your work. 👍

**Comment by mmcc007** [comment]
@fkorotkov We don't currently use tags. But, as far as disabling building on PR's, it looks like it is possible in Travis:
https://docs.travis-ci.com/user/web-ui/#build-pushed-pull-requests

My experimental repo is not mergeable (not a fork), so might be easier if you cut and paste (or something) from:
https://github.com/mmcc007/flutter_arch_samples_cirrus_ci/blob/fac8375a42c3863620dbca2dee65edc3eb5e1cac/.cirrus.yml
...to your version in this PR

**Comment by fkorotkov** [comment]
Oh. I made a typo:

```yaml
only_if: $CIRRUS_PR == ''
```

to **not** run Cirrus for PRs. 😅 I'll add it as well when copying. Sounds good?

**Comment by mmcc007** [comment]
@fkorotkov Yes, I think adding the 'disable on PR' makes sense. That way it only runs when merging to master where we can deal with any (hopefully) transient (Flutter-related) build issues and avoid any potential confusion to contributors. 

Plus it is now possible to turn off PR builds for either travis or cirrus or both.

One good thing is that, after building on Cirrus as well as Travis, we now know that the Flutter timeouts are likely not a function of the build environment and more likely a function of the recent addition of new 'logging-in-place-of-timeouts' feature in Flutter.

**Comment by mmcc007** [comment]
@fkorotkov actually, since cirrus supports it, can you set it to allow only the unit test task to run on PRs for now? That task always runs consistently and does not have the timeout issue. Might disable it too at some later point if overkill.

**Comment by fkorotkov** [comment]
@mmcc007 addressed comments but since the Cirrus app is not installed it's not showing up here yet. 😔😅

**Comment by brianegan** [comment]
@fkorotkov Can I help there? Do I just need to enable Cirrus for this project?

**Comment by brianegan** [comment]
Just installed the app on this project... nothing is showing up in the Cirrus-ci dashboard just yet under "Active Repositories", but it is configured on Github.

Hopefully this means the app is installed and just needs a first run to push data to cirrus?

**Comment by fkorotkov** [comment]
Here is the run for the latest push: https://cirrus-ci.com/build/5681075727630336

**Comment by fkorotkov** [comment]
@brianegan there is a lazy initialization. No need to enable anything. Let me push something so it will appear here.

**Comment by fkorotkov** [comment]
@brianegan you can see it now 👇

**Comment by brianegan** [comment]
Woohoo, thanks! Dang, 28 minutes for the last run is pretty darn fast. I'd say if it's consistently that much faster than Travis it might make sense to move over to it for PRs. What do you think @mmcc007?

**Comment by fkorotkov** [comment]
@brianegan note that 28 minutes is an execution time (11 minutes for unit tests and 17 for the longest integration test). Because of [limits for free runs](https://cirrus-ci.org/faq/#are-there-any-limits) it actually tales longer time. It's like *Run time* vs *Total time* on Travis. Probably should add something like *Total time* indication on Cirrus too... 🤔

Total time on Cirrus will be limited by **all** Integration Tests on macOS which are linear and takes 50 minutes total. But because of [`skip` feature](https://cirrus-ci.org/guide/writing-tasks/#supported-functions) you'll won't run all of them all the time.

**Comment by mmcc007** [comment]
@brianegan With the current runner combinations, 8 linux +1 macOS on Cirrus v's 5 total with max 2 macOS on Travis, Cirrus is currently faster. Travis: ~ 2 hours, Cirrus: ~ 1.4 hours

Cirrus wall clock estimate:

|Unit Test|Linux|macOS|Fudge|Total|
|:-:|-|-|-|-|
|11:47|(longest) 16:49|48:07|10:00|01:21:41|

Cirrus and Travis are both running on similar metal for linux? (both are on Google Cloud).  AFAIK they both use different macOS metal (and possibly different hypervisors?). In general, should probably expect performance on metal to be close from both Cirrus and Travis.

It's possible to use Cirrus's Flutter linux docker image (or equivalent) on Travis for faster installation. So Travis, with it's 2 macOS runners (compared to Cirrus's 1 macOS runner, which is it's performance limiter for us), may possibly still build faster than Cirrus (though it could be close).  

Plus there may be ways to shorten linux jobs on travis by starting the emulator faster which would further narrow the gap. (I've been looking into faster ways to start the emulator)

Metal (and hypervisors?) can change in the future too as new offerings come online and as Travis and Cirrus pick them up (For example, I hear AWS is experimenting with macOS).

As far as what to run, for now, I would run both on master and pick one to run just the unit tests on PR's. Integration tests are the same for all arches, and AFAIK, when merging, 2 full builds will run and the merge will fail if one or both of the full builds fails. This would give contributors much faster turn around time on new code.

So to move forward with this, for now, you could disable builds on PR's for travis in the console and merge this PR... 

After we have a feel for how they both behave, we can pick a preference to run on master. We can always continue to run both if they are similar, and can continue to compare in other ways if we want to change preference.

BTW: there are other ways to go, and I'm ok with whatever direction you take.

@fkorotkov I think Travis reports wall clock time, so yes, Cirrus should probably do something similar. In general, the most relevant build measure is how long a build takes.
 
BTW: Travis auto-starts a default simulator (iPhone X... takes a few seconds to startup). If Cirrus did the same, the following step could be removed:
```yaml
  simulator_script:
    - xcrun simctl list devicetypes
    - xcrun simctl list runtimes
    # create simulator
    - udid=$(xcrun simctl create "iPhone X" com.apple.CoreSimulator.SimDeviceType.iPhone-X com.apple.CoreSimulator.SimRuntime.iOS-12-1)
    # boot simulator
    - xcrun simctl boot $udid
```

(Sorry for the wordy post 🤓😂👍)

**Comment by fkorotkov** [comment]
@mmcc007 sounds good. Could you please [create a new feature request](https://github.com/cirruslabs/cirrus-ci-docs/issues/new?assignees=&labels=feature&template=feature_request.md) for the total time? it might be a little tricky with re-runs though...

Also it's worth to notice that Cirrus enforces limits on per-user basis. So if 3 people working on the same project you'll be able to run total of 24 Linux and 3 macOS tasks in parallel.

**Comment by mmcc007** [comment]
@brianegan This PR LGTM for merging as-is. 
Travis is currently set to run a full build on any new PR or update to a PR. Do you want to disable all builds on PR's for travis in the travis console before merging? (Cirrus is currently set to run the unit tests on any new PR, or an update to a PR)

@fkorotkov Does increasing the #cpus and amount of RAM make any difference for open source projects or does it use a default?

**Comment by fkorotkov** [comment]
@mmcc007 You can experiment with more CPUs but I didn't see much of a difference.

**Comment by mmcc007** [comment]
@fkorotkov what is the max available #CPUs and RAM for open source? What is the default?

**Comment by mmcc007** [comment]
Ah, ic default is 2 CPUs and 4 Gb of memory and max is 8.0 CPUs and up to 24 Gb of memory. Might be worth increasing at some future point.

**Comment by fkorotkov** [comment]
@mmcc007 you got the limits right but note that the more resources the longer it be scheduling (longer to find available resource). So please check beforehand if it actually makes a difference and the build can benefit from multi threading. ☝️

**Comment by mmcc007** [comment]
@fkorotkov FYI: turns out, in practice, with multiple linux runners, the integration tests don't really depend on the unit tests. So, if you want, you can remove the 'depends_on'.

This should improve Cirrus build time by about 15 minutes (the time to run the unit tests)

@brianegan FYI: Will make the corresponding change for Travis builds in a separate PR. This should improve Travis build times by around 30 mins (allows 3 linux runners to run for the first 15 minutes instead of 1).

For us, the # of linux runners is currently a performance limiter on Travis, whereas, the # of macOS runners is a performance limiter on Cirrus. So removing the 'depends_on' in both narrows the gap by about 15 mins. (if I am correct 😂)

**Comment by mmcc007** [comment]
Looks like a potential fix for what looks like the main reason for the  build timeouts is making it's way out:
https://github.com/dart-lang/sdk/issues/33838
https://travis-ci.org/brianegan/flutter_architecture_samples/jobs/509767114#L1435 (linux)
https://travis-ci.org/brianegan/flutter_architecture_samples/jobs/509767094#L228 (macOS)
(the 'optimized out' error shows-up frequently, is not reproducible, and leads to retries and slower or failing builds)

@fkorotkov how often to you upgrade the flutter containers? Can we make occasional requests for updates like this, when they become available?

**Comment by fkorotkov** [comment]
@mmcc007 usually update them the same day with a new stable release. Linux containers are build from [here](https://github.com/cirruslabs/docker-images-flutter) and there is a [separate repo](https://github.com/cirruslabs/osx-images) for all macOS VMs.

**Comment by brianegan** [comment]
@mmcc007 @fkorotkov Excellent, thanks again for trying this out and getting it all setup! Really interesting findings.

 I've just disabled travis on on PRs / pushed branches and I'll merge this one in!
